pkg/ddevapp/config_merge.go

```diff
@@ -0,0 +1,206 @@
+package ddevapp
+
+import (
+	"errors"
+	"log"
+	"regexp"
+
+	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v2"
+)
+
+// mergeConfigToApp does an unmarshall with merging
+func (app *DdevApp) mergeConfigToApp(source []byte) error {
+	// save away state before merges.
+	unmergedApp := &DdevApp{}
+	err := mergo.Merge(unmergedApp, app)
+	if err != nil {
+		return err
+	}
+
+	type mergeData struct {
+		newData interface{}
+		oldData interface{}
+	}
+
+	// get the updated settings. Note that we will replace
+	// anything else from the upstream config for any
+	// key except for web_environment.
+	err = yaml.Unmarshal(source, app)
+	if err != nil {
+		return err
+	}
+
+	// add merges here.  Default strategy is to clobber old
+	// keys.
+	mergeableItems := map[string]mergeData{
+		"web_environment": {
+			&app.WebEnvironment,
+			unmergedApp.WebEnvironment,
+		},
+		"additional_hostnames": {
+			&app.AdditionalHostnames,
+			unmergedApp.AdditionalHostnames,
+		},
+		"additional_fqdns": {
+			&app.AdditionalFQDNs,
+			unmergedApp.AdditionalFQDNs,
+		},
+		"dbimage_extra_packages": {
+			&app.DBImageExtraPackages,
+			unmergedApp.DBImageExtraPackages,
+		},
+		"omit_containers": {
+			&app.OmitContainers,
+			unmergedApp.OmitContainers,
+		},
+		"webimage_extra_packages": {
+			&app.WebImageExtraPackages,
+			unmergedApp.WebImageExtraPackages,
+		},
+		"hooks": {
+			&app.Hooks,
+			unmergedApp.Hooks,
+		},
+	}
+
+	// loop over the items we know how to merge.
+	for item, data := range mergeableItems {
+		switch item {
+		case "web_environment":
+			err = app.mergeWebEnvironment(data.newData, data.oldData.([]string))
+		case "hooks":
+			// merge w/o replacement
+			oldHookData := data.oldData.(map[string][]YAMLTask)
+			err = app.mergeHooks(data.newData, oldHookData)
+		default:
+			// default case is a simple string list merge
+			err = app.mergeStringList(data.newData, data.oldData.([]string))
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// merge an arbitrary string list.
+func (app *DdevApp) mergeStringList(ptr interface{}, oldList []string) error {
+	results := []string{}
+	results = append(results, oldList...)
+
+	newList, ok := ptr.(*[]string)
+	if !ok {
+		return errors.New("unexpected type for DdevApp item")
+	}
+
+	re, _ := regexp.Compile(`^\s*(!*)\s*(\S+)\s*$`)
+
+	// support for a future delete syntax. This stiff runs, but
+	// is not yet used
+	for _, inItem := range *newList {
+		matches := re.FindStringSubmatch(inItem)
+		if matches != nil {
+			if matches[1] != "!" {
+				results = append(results, matches[2])
+			}
+		} else {
+			log.Println("found an invalid string")
+		}
+	}
+
+	// save back the merged results
+	*newList = results
+	return nil
+}
+
+// merge the web_environment string list.
+func (app *DdevApp) mergeWebEnvironment(ptr interface{}, oldEnv []string) error {
```

Comment:
Don't think this is needed?

Reply:
Since this one is called only once, I could do this w/o redirection.  It's most valuable for the string lists; I'd need separate but very similar routines for each and every string list w/o it.  The redirection makes the processing logic a lot more compact.

Also, if we ever do tackle deletion: all that code would need to be touched. I'd as soon do it once, rather than modify very similar code 5 or 6 times.